cluster-api/machines-api/proposal.md

```diff
@@ -0,0 +1,164 @@
+Minimalistic Machines API
+=========================
+
+This proposal is for a minimalistic start to a new Machines API, as part of the
+overall Cluster API project. It is intended to live outside of core Kubernetes
+and add optional machine management features to Kubernetes clusters.
+
+## Capabilities
+
+This API strives to be able to add these capabilities:
+
+1. A new Node can be created in a declarative way, including Kubernetes version
+   and container runtime version. It should also be able to specify
+   provider-specific information such as OS image, instance type, disk
+   configuration, etc., though this will not be portable.
+
+2. A specific Node can be deleted, freeing external resources associated with
+   it.
+
+3. A specific Node can have its kubelet version upgraded or downgraded in a
+   declarative way\*.
+
+4. A specific Node can have its container runtime changed, or its version
+   upgraded or downgraded, in a declarative way\*.
+
+5. A specific Node can have its OS image upgraded or downgraded in a declarative
+   way\*.
+
+\*  It is an implementation detail of the provider if these operations are
+performed in-place or via Node replacement.
+
+## Proposal
+
+This proposal introduces a new API type: Machine. See the full definition in
+[types.go](types.go).
+
+A "Machine" is the declarative spec for a Node, as represented in Kuberenetes
+core. If a new Machine object is created, a provider-specific controller will
+handle provisioning and installing a new host to register as a new Node matching
+the Machine spec. If the Machine's spec is updated, a provider-specific
+controller is responsible for updating the Node in-place or replacing the host
+with a new one matching the updated spec. If a Machine object is deleted, the
+corresponding Node should have its external resources released by the
+provider-specific controller, and should be deleted as well.
+
+Fields like the kubelet version, the container runtime to use, and its version,
+are modeled as fields on the Machine's spec. Any other information that is
+provider-specific, though, is part of an opaque ProviderConfig string that is
+not portable between different providers.
+
+The ProviderConfig is recommended to be a serialized API object in a format
+owned by that provider, akin to the [Component Config](https://goo.gl/opSc2o)
+pattern. This will allow the configuration to be strongly typed, versioned, and
+have as much nested depth as appropriate. These provider-specific API
+definitions are meant to live outside of the Machines API, which will allow them
+to evolve independently of it. Attributes like instance type, which network to
+use, and the OS image all belong in the ProviderConfig.
+
+## In-place vs. Replace
+
+One simplification that might be controversial in this proposal is the lack of
+API control over "in-place" versus "replace" reconciliation strategies. For
```

Comment:
Users may end up wanting this to make a trade-off between the disruptiveness and cleanliness of a rollout, but I think it's fine to push this down to the ProviderConfig and leave it out of the top-level API.